pyannote/audio/generators/change.py

```diff
@@ -0,0 +1,130 @@
+from scipy.stats import zscore
+import numpy as np
+
+
+class ChangeDetectionBatchGenerator(PeriodicFeaturesMixin,
+                                            FileBasedBatchGenerator):
+
+    """(X_batch, y_batch) batch generator
+
+    Yields batches made of subsequences obtained using a sliding
+    window over the audio files.
+
+    Parameters
+    ----------
+    feature_extractor : YaafeFeatureExtractor
+    duration: float, optional
+        yield semgents of length `duration`
+        Default to 3.2s.
+    step: float, optional
+        Duration and step of sliding window (in seconds).
+        Default to 0.8s.
+    balance: float, optional
+        Frames within a context window centred by change 
+        point will be labeled by 1. The length of the context 
+        window is 2*balance.
+        Default to 0.01s.
+    batch_size: int, optional
+        Size of batch
+
+    Returns
+    -------
+    X_batch : (batch_size, n_samples, n_features) numpy array
+        Batch of feature sequences
+    y_batch : (batch_size, n_samples) numpy array
+        Batch of corresponding label sequences
+
+    Usage
+    -----
+    >>> batch_generator = ChangeDetectionBatchGenerator(feature_extractor)
+    >>> for X_batch, y_batch in batch_generator.from_file(current_file):
+    ...     # do something with
+    """
+
+    def __init__(self, feature_extractor,
+                 balance=0.01, duration=3.2, step=0.8, batch_size=32):
+
+        self.feature_extractor = feature_extractor
+        self.duration = duration
+        self.step = step
+        self.balance = balance
+
+        segment_generator = SlidingSegments(duration=duration,
+                                            step=step,
+                                            source='annotated')
+        super(ChangeDetectionBatchGenerator, self).__init__(
+            segment_generator, batch_size=batch_size)
+
+    def signature(self):
+
+        shape = self.shape
+
+        return [
+            {'type': 'sequence', 'shape': shape},
+            {'type': 'sequence', 'shape': (shape[0], 2)}
+        ]
+
+    def preprocess(self, current_file, identifier=None):
+        """Pre-compute file-wise X and y"""
+
+        # extract features for the whole file
+        # (if it has not been done already)
+        current_file = self.periodic_preprocess(
+            current_file, identifier=identifier)
+
+        # if labels have already been extracted, do nothing
+        if identifier in self.preprocessed_.setdefault('y', {}):
+            return current_file
+
+        # get features as pyannote.core.SlidingWindowFeature instance
+        X = self.preprocessed_['X'][identifier]
+        sw = X.sliding_window
+        n_samples = X.getNumber()
+
+        y = np.zeros((n_samples + 4, 1), dtype=np.int8)-1
+        # [-1] ==> unknown / [0] ==> not change part / [1] ==> change part
+
+        annotated = current_file.get('annotated', X.getExtent())
```

Comment:
There is now a function called `get_annotated` in `pyannote.database.utils` that you should be using instead.